include/dr/mhp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,275 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#ifdef SYCL_LANGUAGE_VERSION
+#include <oneapi/dpl/async>
+#endif
+
+namespace dr::mhp {
+
+namespace __detail {
+
+template <dr::distributed_contiguous_range R, dr::distributed_iterator O,
+          typename BinaryOp, typename U = rng::range_value_t<R>>
+auto inclusive_scan_impl_(R &&r, O &&d_first, BinaryOp &&binary_op,
+                          std::optional<U> init = {}) {
+  assert(aligned(r, d_first));
+
+  auto d_last = d_first;
+  rng::advance(d_last, rng::size(r));
+
+  // TODO: current alg assumes local segments are contiguous and located
+  //  in rank order. This is OK as all data structures we have currently have
+  //  just one segment per rank and in rank order.
+  auto in_segments = local_segments(r);
+  auto out_segments = local_segments(rng::subrange(d_first, d_last));
+
+  // rng::size(in_segments) doesn't work and can't count non-empty segments
+  std::size_t local_segments_count = 0;
+  for (auto &&s [[maybe_unused]] : in_segments)
+    if (!rng::empty(s))
+      ++local_segments_count;
+
+#ifdef SYCL_LANGUAGE_VERSION
+  std::vector<sycl::event> events;
+#endif
+
+  using OVal = rng::iter_value_t<O>;
+  // how to get allocator used with input and output range? I assume it a
+  // default one probably we should forbid using non-default ones with
+  // distributed_ranges containers (like distributed_vector)
+  OVal *local_partial_sums =
```

Comment:
`default_allocator` allocates on the device. You are performing the operation on the host so you need std::allocator. But even better to just use a std::vector so the destructor will take care of deallocation.

Reply:
This operation is performed on device. But I've changed raw pointer to `unique_ptr` with deleter, which solves deallocation code smell that was here. I've added a description in comment which lists steps and I've marked what is offloaded.